movement/lib/TOTP/sha256.c

```diff
@@ -0,0 +1,372 @@
+    mbedtls_sha256_update( ctx, msglen, 8 );
+
+    PUT_UINT32_BE( ctx->state[0], output,  0 );
+    PUT_UINT32_BE( ctx->state[1], output,  4 );
+    PUT_UINT32_BE( ctx->state[2], output,  8 );
+    PUT_UINT32_BE( ctx->state[3], output, 12 );
+    PUT_UINT32_BE( ctx->state[4], output, 16 );
+    PUT_UINT32_BE( ctx->state[5], output, 20 );
+    PUT_UINT32_BE( ctx->state[6], output, 24 );
+
+    if( ctx->is224 == 0 )
+        PUT_UINT32_BE( ctx->state[7], output, 28 );
+}
+
+/*
+ * output = SHA-256( input buffer )
+ */
+void mbedtls_sha256( const unsigned char *input, size_t ilen,
+             unsigned char* output, int is224 )
+{
+    mbedtls_sha256_context ctx;
+
+    mbedtls_sha256_init( &ctx );
+    mbedtls_sha256_starts( &ctx, is224 );
+    mbedtls_sha256_update( &ctx, input, ilen );
+    mbedtls_sha256_finish( &ctx, output );
+    mbedtls_sha256_free( &ctx );
+}
+
+/*
+* Compute HMAC_SHA224/256 using key, key length, text to hash, size of the text, output buffer and a switch for SHA224
+*/
+void HMAC_SHA256(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, uint8_t* out, int is224){
+  int digest_length = SHA256_DIGEST_LENGTH;
+  if (is224 == 1) {
+    digest_length = SHA224_DIGEST_LENGTH;
+  }
+  
+  uint8_t i;
+  uint8_t k_ipad[SHA256_BLOCK_LENGTH]; /* inner padding - key XORd with ipad */
+  uint8_t k_opad[SHA256_BLOCK_LENGTH]; /* outer padding - key XORd with opad */
+  uint8_t buffer[SHA256_BLOCK_LENGTH + digest_length];
+
+  /* start out by storing key in pads */
+  memset(k_ipad, 0, sizeof(k_ipad));
+  memset(k_opad, 0, sizeof(k_opad));
+
+  if (key_length <= SHA256_BLOCK_LENGTH) {
+      memcpy(k_ipad, key, key_length);
+      memcpy(k_opad, key, key_length);
+  }
+
+  else {
+      mbedtls_sha256(key, key_length, k_ipad, is224);
+      memcpy(k_opad, k_ipad, SHA256_BLOCK_LENGTH);
+  }
+
+  /* XOR key with ipad and opad values */
+  for (i = 0; i < SHA256_BLOCK_LENGTH; i++) {
+      k_ipad[i] ^= HMAC_IPAD;
+      k_opad[i] ^= HMAC_OPAD;
+  }
+  
+  // perform inner SHA256
+  memcpy(buffer, k_ipad, SHA256_BLOCK_LENGTH);
+  memcpy(buffer + SHA256_BLOCK_LENGTH, in, n);
+  mbedtls_sha256(buffer, SHA256_BLOCK_LENGTH + n, out, is224);
+  
+  memset(buffer, 0, SHA256_BLOCK_LENGTH + n);
+
+  // perform outer SHA256
+  memcpy(buffer, k_opad, SHA256_BLOCK_LENGTH);
+  memcpy(buffer + SHA256_BLOCK_LENGTH, out, digest_length);
+  mbedtls_sha256(buffer, SHA256_BLOCK_LENGTH + digest_length, out, is224);
+}
+
+/*
+* Compute TOTP_HMAC_SHA224/256 using key, key length, text to hash, size of the text and a switch for SHA224
+*/
+uint32_t TOTP_HMAC_SHA256(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, int is224){
+    int digest_length = SHA256_DIGEST_LENGTH;
+    if (is224 == 1) {
+      digest_length = SHA224_DIGEST_LENGTH;
+    }
+
+    // STEP 1, get the HMAC-SHA256 hash from counter and key
+    uint8_t hash[digest_length];
+    HMAC_SHA256(key, key_length, in, 8, hash, is224);

```

Comment:
Same issue as in TOTP_HMAC_SHA1.

Reply:
You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.